### PR TITLE
Net 846 add upwards scaling migration

### PR DESCRIPTION
**Pull request by ptesavol**
* Data migration works in the case of growing network topology
* Cleaning up of connections when leaving works

### CODE REVIEW

File: packages/dht/test/integration/MigrateData.test.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        })
+
+        const closestNode = nodesById.get(PeerID.fromValue(closest[0].getPeerDescriptor().kademliaId).toKey())!
+
+        // @ts-expect-error private field
+        expect(closestNode.localDataStore.getEntry(dataKey)).toBeTruthy()
+    }, 180000)
+
+    /*
+    it.only('Data migrates to the last remaining node if all other nodes leave gracefully', async () => {
```
Comment (juslesan) [inline packages/dht/test/integration/MigrateData.test.ts:158]: Is this test case intentionally commented out?

File: packages/dht/src/connection/ManagedConnection.ts [hunk 2]
```diff
@@ -245,9 +273,16 @@ export class ManagedConnection extends EventEmitter<Events> {
...
-                        ['bufferSentByOtherConnection', 'closing'], 15000)
-
+                    let result2: RunAndRaceEventsReturnType<Events>
+
+                    try {
+                        result2 = await raceEvents3<Events>(this,
+                            ['bufferSentByOtherConnection', 'closing'], 15000)
+                    } catch (ex) {
+                        logger.error('Exception from raceEvents3 while waiting bufferSentByOtherConnection or closing ' + ex)
+                        //const dataString = protoToString(Message.fromBinary(data), Message)
```
Comment (juslesan) [inline packages/dht/src/connection/ManagedConnection.ts:283]: remove commented code

File: packages/dht/src/connection/ManagedConnection.ts [hunk 3]
```diff
@@ -230,8 +250,16 @@ export class ManagedConnection extends EventEmitter<Events> {
...
-            const result = await raceEvents3<Events>(this, ['handshakeCompleted', 'handshakeFailed',
-                'bufferSentByOtherConnection', 'closing', 'disconnected'], 15000)
+            let result: RunAndRaceEventsReturnType<Events>
+
+            try {
+                result = await raceEvents3<Events>(this, ['handshakeCompleted', 'handshakeFailed',
+                    'bufferSentByOtherConnection', 'closing', 'disconnected'], 15000)
+            } catch (e) {
+                logger.error('Exception from raceEvents3 ' + e)
+                //const dataString = protoToString(Message.fromBinary(data), Message)
```
Comment (juslesan) [inline packages/dht/src/connection/ManagedConnection.ts:260]: remove commented code

File: packages/dht/src/connection/ManagedConnection.ts [hunk 4]
```diff
@@ -214,12 +217,29 @@ export class ManagedConnection extends EventEmitter<Events> {
...
     }
 
-    private onDisconnected(code?: number, reason?: string): void {
-        logger.trace('onDisconnected ' + code + ' ' + reason)
-        this.doDisconnect()
+    private onDisconnected(disconnectionType: DisconnectionType, _code?: number, _reason?: string): void {
+        
+        logger.trace(' ' + this.ownPeerDescriptor.nodeName + ', ' + this.peerDescriptor?.nodeName + ' onDisconnected() ' + disconnectionType )
+        if (this.bufferSentbyOtherConnection) {
+            //logger.error('onDisconnected called on replaced connection')
```
Comment (juslesan) [inline packages/dht/src/connection/ManagedConnection.ts:224]: remove commented code

File: packages/dht/src/connection/ConnectionManager.ts [hunk 5]
```diff
@@ -384,20 +389,26 @@ export class ConnectionManager extends EventEmitter<Events> implements ITranspor
         this.onConnectionCountChange()
     }
 
-    private onDisconnected = (connection: ManagedConnection) => {
-        if (!this.started || this.stopped) {
-            return
-        }
-        logger.trace(this.config.nodeName + ' onDisconnected()')
+    private onDisconnected = (connection: ManagedConnection, disconnectionType: DisconnectionType) => {
+        logger.trace(' ' + this.config.nodeName + ', ' + connection.getPeerDescriptor()?.nodeName +
```
Comment (juslesan) [inline packages/dht/src/connection/ConnectionManager.ts:393]: Logging starts with `'  '`

Comment (juslesan) [reply packages/dht/src/connection/ConnectionManager.ts:393]: Also on lines 401, 406

File: packages/dht/src/dht/store/DataStore.ts [hunk 6]
```diff
@@ -34,8 +45,88 @@ export class DataStore implements IStoreService {
...
+        this.storeData = this.storeData.bind(this)
+        this.migrateData = this.migrateData.bind(this)
+        this.config.rpcCommunicator!.registerRpcMethod(StoreDataRequest, StoreDataResponse, 'storeData', this.storeData)
+        this.config.rpcCommunicator!.registerRpcMethod(MigrateDataRequest, MigrateDataResponse, 'migrateData', this.migrateData)
+
+        this.config.dhtNodeEmitter.on('newContact', (peerDescriptor: PeerDescriptor, _closestPeers: PeerDescriptor[]) => {
+
+            this.config.localDataStore.store.forEach((dataMap, _dataKey) => {
+                dataMap.forEach((dataEntry) => {
+                    //if (this.isFurtherFromDataThan(dataEntry, contact) &&
```
Comment (juslesan) [inline packages/dht/src/dht/store/DataStore.ts:57]: Remove commented code blocks from this file. Lines starting from 57, 86, 200, 210,

File: packages/dht/src/dht/store/DataStore.ts [hunk 7]
```diff
@@ -76,9 +170,84 @@ export class DataStore implements IStoreService {
...
+        })
+
+        /*
+        if (!sortedList.getAllContacts()[0].peerId.equals(this.ownPeerId!)) {
+            // If we are not the closest node to the data, do not migrate
+            return false
+        }
+        */
+
+        if (!sortedList.getAllContacts()[0].getPeerId().equals(ownPeerId)) {
```
Comment (juslesan) [inline packages/dht/src/dht/store/DataStore.ts:217]: Could split the cases under this if else block to their own methods such as  `migrateToClosestNode` and `migrateToAllClosestNodes`

File: packages/dht/src/connection/WebRTC/NodeWebRtcConnection.ts [hunk 8]
```diff
@@ -81,7 +82,7 @@ export class NodeWebRtcConnection extends EventEmitter<Events> implements IConne
 
         this.connectingTimeoutRef = setTimeout(() => {
             logger.trace('connectingTimeout, this.closed === ' + this.closed)
-            this.doClose()
+            this.doClose('OTHER')
```
Comment (juslesan) [inline packages/dht/src/connection/WebRTC/NodeWebRtcConnection.ts:85]: Disconnection case OTHER is set in quite many disconnect cases. What is it used to designate?

**Review by juslesan — CHANGES_REQUESTED** [review]
A few code cleanup related comments to address before merging.

File: packages/dht/test/integration/MigrateData.test.ts (hunk 1 above)
Comment (ptesavol) [reply packages/dht/test/integration/MigrateData.test.ts:158]: yes, this is the downwards migration part that will be enabled in next pr

File: packages/dht/src/connection/ConnectionManager.ts (hunk 5 above)
Comment (ptesavol) [reply packages/dht/src/connection/ConnectionManager.ts:393]: this is intentional, we can search for ' 2, 4 ' in text editor, and '32, 47' does not match the search

File: packages/dht/src/connection/WebRTC/NodeWebRtcConnection.ts (hunk 8 above)
Comment (ptesavol) [reply packages/dht/src/connection/WebRTC/NodeWebRtcConnection.ts:85]: I think we should split it to 'CONNECTION_TIMEOUT', 'INCOMING_CLOSE' etc. Left it like that for now, and will come back to it when also the downwards migration works.

**Pull request merged**